### PR TITLE
Add Rectanglei#containsPoint(float)

### PR DESCRIPTION
**Pull request by pollend**

### CODE REVIEW

File: src/org/joml/Rectanglei.java [hunk 1]
```diff
@@ -284,6 +284,29 @@ public boolean containsPoint(Vector2ic point) {
...
 
+    /**
+     * Test whether the point <code>(x, y)</code> lies inside this BlockRegion.
+     *
+     * @param x the x coordinate of the point
+     * @param y the y coordinate of the point
+     * @return <code>true</code> iff the given point lies inside this BlockRegion; <code>false</code> otherwise
+     */
+    public boolean containsPoint(float x, float y) {
+        return x >= this.minX && y >= this.minY && x <= this.maxX && y <= this.maxY;
```
Comment (httpdigest) [inline src/org/joml/Rectanglei.java:295]: I wonder whether the upper bounds should rather be: `x < this.maxX` and `y < this.maxY` as current this is inconsistent with the `int` version of this method, yielding for exact integer coordinates false with the `int` version and true for the `float` version (for exactly float-representable integers).

Comment (pollend) [reply src/org/joml/Rectanglei.java:295]: AABBi uses `<=` for the max bound does the edge of a square count as containing a point?

Comment (httpdigest) [reply src/org/joml/Rectanglei.java:295]: Hmm... it seems that making this consistent everywhere and documenting it in the JavaDocs should be the first thing to do. I'm fine with either way, ("upper edge" being inclusive or exclusive). Do you have any preferences that would make your life easier?

Comment (pollend) [reply src/org/joml/Rectanglei.java:295]: if its for consistency then inclusive would be better.

**Comment by pollend** [comment]
I changed the it to include the edge of the bound and wrote some specific test for this case.

**Comment by httpdigest** [comment]
> I changed Rectanglei/f to include the edge of the bound for testing intersectsRectangle and containsPoint and wrote some specific test for this case.

Actually, this also seems weird. Why would a rectangle (0, 0)<(1,1) have an intersection with (1,1)<(2,2)? The first ends at (1,1) and the second starts at (1,1), but they do not overlap. I think this should be thought about more. Maybe let's see how other libraries/languages handle this.

**Comment by pollend** [comment]
I think I'll write some similar test cases for AABB.

**Comment by pollend** [comment]
@httpdigest so I have some test cases for some edge cases to consider. what do you think could be adjusted for the intersection logic?

**Comment by pollend** [comment]
here is an aabb implemented for three.js: https://github.com/mrdoob/three.js/blob/dev/src/math/Box3.js . I updated the methods to line up with this implementation.

**Comment by 4Denthusiast** [comment]
In the Rectanglei case, having both of the bounds be exclusive is inconsistent with `area`, which returns the number of points contained with half-inclusive edges. Also changing this behaviour is likely to break existing uses. The similar change to Rectangled (making its edges exclusive) is breaking, but less so because in most cases reliance on precide floating point equality should be avoided anyway. The exclusive boundary style also has the oddity that some `Rectanglei`s are valid (according to their `isValid` method) but contain no (int) points.

The Box3 methods are actually inclusive, they're just written in a confusing way.

Half-inclusive bounds are the only case which, when implemented consistently, give the size of an int AABB as the same as the corresponding float AABB (i.e. the one containing all the same floats). I'm in favour of using half-open for all the int versions and either closed, open or half-open (but preferably half-open) for the floating point versions.

Part of the reason I want it all to be half-open is that that seems like the most obvious way to do it, which follows from all the other reasons but is pretty subjective, but as an additional argument in favour of half-open being the obvious assumption, that's what you @pollend incorrectly assumed AABBi was using yourself when writing https://github.com/MovingBlocks/Terasology/blob/develop/engine/src/main/java/org/terasology/world/block/BlockRegion.java.

**Comment by pollend** [comment]
The other thing you can do is add a border around the bounds and only use floating point values. For consistency, it would all be close bounded. It would allow for inclusive and exclusive bounds depending on the direction of the border along with how much to include from the border of the bounds. In most cases for all intensive purposes the value used for finding a point with a bounds is negligible. I only really see this problem for the integer case.

**Comment by 4Denthusiast** [comment]
@pollend Do you mean something along the lines of `x >= xMin - ε && x <= xMax + ε`? Why would that be useful? What do you mean by "only use floating point values"?

**Comment by pollend** [comment]
yep, there is less to interpret wrong in this case since the boundary can either be open or closed depending on the direction of the border.

**Comment by 4Denthusiast** [comment]
I assume that in most cases where AABBi (or Rectanglei) is used, it will be the set of int vectors it contains that's most important, like block regions in Terasology, so the way those classes work should be optimised to make most sense in that case. For the scenario you describe, representing a subset of R^3 whose boundaries happen to always have integer coordinates, just using an AABBf or AABBd with integer values seems more appropriate.

**Comment by pollend** [comment]
Maybe AABBi and Rectanglei is not really that useful since the associated case is kind of specific. Might make sense to depercate both classes?

**Comment by 4Denthusiast** [comment]
I don't know what else this library is used for than Terasology, so I don't know how generally applicable those classes are, but having AABBi and AABBf both defined in the same place would at least be more consistent. Also, even within Terasology, AABBis are used in 2 sort-of separate ways: for sets of block positions and sets of chunks.

**Comment by httpdigest** [comment]
Sorry, for taking so long right now. I'm finding that deciding on any particular convention for AABB operations, such as whether a point lying on an edge is to be considered inside or outside of the AABB, is really tricky, and as mentioned before by others, really use-case dependent.
I am going to merge the JOML PRs related to this for your work on Terasology for the sake of you being able to continue work on Terasology. But I am very inclined to completely drop support for the AABB, Circle, Rectangle (and basically all geometric shapes) classes soon. It _is_ really use-case dependent on how one would go about defining them both in terms of their operations' semantics as well as their way of storing information (e.g. AABB = min/max corner or center and half sizes, for example).

**Comment by pollend** [comment]
that makes sense.
